libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -3363,16 +3376,26 @@ void IRGeneratorForStatements::revertWithError(
 	vector<ASTPointer<Expression const>> const& _errorArguments
 )
 {
+	bool needsAllocation = true;
+	if (optional<size_t> size = staticEncodingSize(_parameterTypes))
+		if (ranges::all_of(_parameterTypes, [](auto const* type) { return type && type->isValueType(); }))
+			needsAllocation = *size + 4 > CompilerUtils::generalPurposeMemoryStart;
```

Comment:
So, just to make sure I understand why this is memory safe:
- It's ok to overwrite free mem pointer because we're reverting and not allocating anything before then.
- It's ok to overwrite the zero slot since the condition above excludes any empty dynamic arrays passed in as arguments.
- It's not ok to overwrite the area where variables moved to memory might be because one of them might be `<pos>` or `<end>`.

Is that correct?

Reply:
Yes, all that's true - but the restriction of types is in the current version not strong enough to exclude all uses of the scratch space during the encoding step, so that still needs fixing (restricting to value types for both the types of the arguments *and* the types of the error signature should do that - pending a closer look, though).